src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,532 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// The string returned from this function represents the inputs to the compiler which impact determinism.  It is 
+    /// meant to be inline with the specification here:
+    /// 
+    ///     - https://github.com/dotnet/roslyn/blob/main/docs/compilers/Deterministic%20Inputs.md
+    /// 
+    /// Options which can cause compilation failure, but doesn't impact the result of a successful
+    /// compilation should be included. That is because it is interesting to describe error states
+    /// not just success states. Think about caching build failures as well as build successes.
+    ///
+    /// When an option is omitted, say if there is no value for a public crypto key, we should emit
+    /// the property with a null value vs. omitting the property. Either approach would produce 
+    /// correct results the preference is to be declarative that an option is omitted.
+    /// </summary>
+    internal abstract class DeterministicKeyBuilder
+    {
+        protected DeterministicKeyBuilder()
+        {
+        }
+
+        protected void WriteFilePath(
+            JsonWriter writer,
+            string propertyName,
+            string? filePath,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions options)
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
+            }
+            else if (filePath is not null)
+            {
+                filePath = PathUtilities.NormalizePathPrefix(filePath, pathMap);
+            }
+
+            writer.Write(propertyName, filePath);
+        }
+
+        internal static string EncodeByteArrayValue(ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            return builder.ToStringAndFree();
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
+        {
+            foreach (var b in value)
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
+
+        protected static void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
```

Comment:
method could be simplified to `writer.Write(name, EncodeByteArrayValue(value))`, avoiding duplication with above `EncodeByteArrayValue`?

Reply:
Doing that though would introduce an unnecessary intermediate `string` allocation.